libraries/chain/apply_context.cpp

```diff
@@ -188,6 +165,55 @@ void apply_context::exec_one()
    }
 }
 
+void apply_context::validate_account_ram_deltas() {
+   const size_t checktime_interval = 10;
+   size_t counter = 0;
+   bool not_in_notify_context = (receiver == act->account);
+   const auto end = _account_ram_deltas.end();
+   for( auto itr = _account_ram_deltas.begin(); itr != end; ++itr, ++counter ) {
+      // wlog("pending RAM delta: ${account} ${delta}", ("account", itr->account)("delta", itr->delta) );
+      if( counter == checktime_interval ) {
+         trx_context.checktime();
+         counter = 0;
+      }
+      if( !privileged && itr->delta > 0 && itr->account != receiver ) {
+         SYS_ASSERT( not_in_notify_context, unauthorized_ram_usage_increase,
+                     "unprivileged contract cannot increase RAM usage of another account within a notify context: ${account}",
+                     ("account", itr->account)
+         );
+         SYS_ASSERT( has_authorization( itr->account ), unauthorized_ram_usage_increase,
+                     "unprivileged contract cannot increase RAM usage of another account that has not authorized the action: ${account}",
+                     ("account", itr->account)
+         );
+      }
+      auto& payer = itr->account;
+      auto& ram_delta = itr->delta;
+      if (payer != receiver && ram_delta > 0) {
+         if (receiver == config::system_account_name) {
+         } else if (payer == config::system_account_name && is_privileged()) {
+         } else {
+            auto payer_found = false;
+            // search act->authorization for a payer permission role
+            for( const auto& auth : act->authorization ) {
+               if( auth.actor == payer && auth.permission == config::sysio_payer_name ) {
+                  payer_found = true;
+                  break;
+               }
+               if ( auth.actor == config::system_account_name ) {
+                  // If the payer is the system account, we don't enforce explicit payer authorization.
+                  // This avoids a lot of changes for tests and sysio should not be calling untrusted contracts
+                  // or spending user's RAM unexpectedly.
+                  payer_found = true;
+                  break;
```

Comment:
optimization opportunity, no need to do the auth loop for additional deltas.  but maybe the explicit flow is better, so feel free to leave as-is